scipy/spatial/transform/rotation.py

```diff
@@ -0,0 +1,862 @@
+    inv
+    __len__
+    __getitem__
+    """
+    def __init__(self, quat, normalized=False, copy=True):
+        self._single = False
+        # Try to convert to numpy array
+        quat = np.asarray(quat, dtype=float)
+
+        if quat.ndim not in [1, 2] or quat.shape[-1] != 4:
+            raise ValueError("Expected `quat` to have shape (4,) or (N x 4), "
+                             "got {}.".format(quat.shape))
+
+        # If a single quaternion is given, convert it to a 2D 1 x 4 matrix but
+        # set self._single to True so that we can return appropriate objects
+        # in the `to_...` methods
+        if quat.shape == (4,):
+            quat = quat[None, :]
+            self._single = True
+
+        if normalized:
+            self._quat = quat.copy() if copy else quat
+        else:
+            self._quat = quat.copy()
+            # L2 norm of each row
+            norms = scipy.linalg.norm(quat, axis=1)
+
+            # Raise ValueError for zero (eps?) norm quaternions
+            zero_norms = norms == 0
+            if zero_norms.any():
+                raise ValueError("Found zero norm quaternions in `quat`.")
+
+            # Normalize each quaternion, ensuring norm is broadcasted along
+            # each column.
+            self._quat[~zero_norms] /= norms[~zero_norms][:, None]
+
+    def __len__(self):
+        """Number of rotations contained in this object.
+
+        Multiple rotations can be stored in a single instance.
+
+        Returns
+        -------
+        length : int
+            Number of rotations stored in object.
+        """
+        return self._quat.shape[0]
+
+    @classmethod
+    def from_quaternion(cls, quat, normalized=False):
+        """Initialize Rotation from quaternions.
+
+        3D rotations can be represented using unit-norm quaternions [1]_.
+
+        Parameters
+        ----------
+        quat : array_like, shape (N, 4) or (4,)
+            Each row is a (possibly non-unit norm) quaternion in scalar-last
+            (x, y, z, w) format.
+        normalized : boolean, optional
+            If `False`, input quaternions are normalized to unit norm before
+            being stored. If `True`, quaternions are assumed to already have
+            unit norm and are stored as given. Default is `False`.
+
+        Returns
+        -------
+        output : `Rotation` instance
+            Object containing the rotations represented by input quaternions.
+
+        References
+        ----------
+        .. [1] `Quaternions and Spatial Rotation
+               <https://en.wikipedia.org/wiki/Quaternions_and_spatial_rotation>`_
+        """
+
+        return cls(quat, normalized)
+
+    def as_quaternion(self):
+        """Represent rotations as quaternions.
+
+        Rotations in 3 dimensions can be represented using unit norm
+        quaternions [1]_. The mapping from quaternions to rotations is
+        two-to-one, i.e. a quaternion and its additive inverse represent the
```

Comment:
Additive inverse? Write something simpler, like "If you change the sign of each component you will get the same rotation." (But better than that, couldn't formulate it quickly).